src/raftstore/store/store.rs

```diff
@@ -1456,6 +1616,241 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+            admin.mut_merge().set_source(region.clone());
+            admin.mut_merge().set_commit(state.get_commit());
+            admin
+                .mut_merge()
+                .set_entries(RepeatedField::from_vec(entries));
+            request.set_admin_request(admin);
+            request
+        };
+        // Please note that, here assumes that the unit of network isolation is store rather than
+        // peer. So a quorum stores of souce region should also be the quorum stores of target
+        // region. Otherwise we need to enable proposal forwarding.
+        self.propose_raft_command(req, Callback::None);
+        Ok(())
+    }
+
+    fn rollback_pre_merge(&mut self, region: &metapb::Region) {
+        let req = {
+            let peer = &self.region_peers[&region.get_id()];
+            let state = peer.pending_merge.as_ref().unwrap();
+            let mut request = new_admin_request(region.get_id(), peer.peer.clone());
+            request
+                .mut_header()
+                .set_region_epoch(peer.region().get_region_epoch().clone());
+            let mut admin = AdminRequest::new();
+            admin.set_cmd_type(AdminCmdType::RollbackPreMerge);
+            admin
+                .mut_rollback_pre_merge()
+                .set_commit(state.get_commit());
+            request.set_admin_request(admin);
+            request
+        };
+        self.propose_raft_command(req, Callback::None);
+    }
+
+    fn on_check_merge(&mut self, event_loop: &mut EventLoop<Self>) {
+        let merge_states = mem::replace(&mut self.merge_states, vec![]);
+        for region in &merge_states {
+            if let Err(e) = self.schedule_merge(region) {
+                info!(
+                    "[region {}] failed to schedule merge, rollback: {:?}",
+                    region.get_id(),
+                    e
+                );
+                self.rollback_pre_merge(region);
+            }
+        }
+        self.merge_states = merge_states;
+        self.register_merge_check_tick(event_loop);
+    }
+
+    fn on_ready_pre_merge(&mut self, region: metapb::Region, state: MergeState) {
+        {
+            let peer = self.region_peers.get_mut(&region.get_id()).unwrap();
+            peer.pending_merge = Some(state);
+            peer.mut_store().region = region.clone();
+        }
+
+        if let Err(e) = self.schedule_merge(&region) {
+            info!(
+                "[region {}] failed to schedule merge, rollback: {:?}",
+                region.get_id(),
+                e
+            );
+            self.rollback_pre_merge(&region);
+        }
+        self.merge_states.push(region);
+    }
+
+    fn on_ready_merge(&mut self, region: metapb::Region, source: metapb::Region) {
+        let source_peer = {
+            let peer = self.region_peers.get_mut(&source.get_id()).unwrap();
+            // In some case, PreMerge entry may be carried by the dst region. So
```

Comment:
It confused me. Any example?

Reply:
That's the same case when entries in `Merge` message is in use.